crates/matrix-sdk-ui/src/authentication/oidc.rs

```diff
@@ -0,0 +1,214 @@
+use std::{
+    collections::{hash_map::DefaultHasher, HashMap},
+    fs,
+    fs::File,
+    hash::{Hash, Hasher},
+    io::{BufReader, BufWriter},
+    path::PathBuf,
+};
+
+use matrix_sdk::oidc::types::registration::VerifiedClientMetadata;
+use serde::{Deserialize, Serialize};
+
+#[derive(Debug, thiserror::Error)]
+pub enum OidcRegistrationsError {
+    #[error("Failed to use the supplied base path.")]
+    InvalidBasePath,
+}
+
+/// The data needed to restore an OpenID Connect session.
+#[derive(Debug)]
+pub struct OidcRegistrations {
+    /// The path of the file where the registrations are stored.
+    file_path: PathBuf,
+    /// The hash for the metadata used to register the client.
+    /// This is used to check if the client needs to be re-registered.
+    metadata_hash: u64,
+    /// Pre-configured registrations for use with issuers that don't support
+    /// dynamic client registration.
+    static_registrations: HashMap<String, String>,
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+struct RegistrationData {
+    /// The hash for the metadata used to register the client.
+    metadata_hash: u64,
+    /// All of the registrations this client has made as a HashMap of issuer URL
+    /// (as a string) to client ID (as a string).
+    dynamic_registrations: HashMap<String, String>,
+}
+
+/// Manages the storage of OIDC registrations.
+impl OidcRegistrations {
+    pub fn new(
+        base_path: &str,
+        metadata: &VerifiedClientMetadata,
+        static_registrations: HashMap<String, String>,
+    ) -> Result<Self, OidcRegistrationsError> {
+        let oidc_directory = PathBuf::from(base_path).join("oidc");
+        fs::create_dir_all(&oidc_directory).map_err(|_| OidcRegistrationsError::InvalidBasePath)?;
+
+        let metadata_hash = {
+            let mut hasher = DefaultHasher::new();
+            let metadata_string = serde_json::to_string(metadata).unwrap();
+            metadata_string.hash(&mut hasher);
+            hasher.finish()
+        };
+
+        Ok(OidcRegistrations {
+            file_path: oidc_directory.join("registrations.json"),
+            metadata_hash,
+            static_registrations,
+        })
+    }
+
+    /// Returns the underlying registration data.
+    fn registration_data(&self) -> RegistrationData {
+        let reader = match File::open(&self.file_path) {
+            Ok(file) => BufReader::new(file),
+            Err(e) => {
+                tracing::warn!("Failed to open registrations file: {e}");
+                return RegistrationData {
+                    metadata_hash: self.metadata_hash,
+                    dynamic_registrations: Default::default(),
+                };
```

Comment:
Since this is an internal function, it would be better to properly bubble up this error to the caller instead (i.e. have this function return a `Result<RegistrationData, ...>`).